src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -72,95 +95,152 @@ public HdfsWriter(HdfsSinkConnectorConfig connectorConfig) {
     }
   }
 
-
   public Map<TopicPartition, Long> getPreviousOffsets() {
     return offsets;
   }
 
-  public void writeRecord(TopicPartition topicPart, SinkRecord record)
-      throws CopycatException{
+  public void writeRecord(final TopicPartition topicPart, final SinkRecord record)
+      throws CopycatException, InterruptedException {
+    lock.lock();
     try {
+      recover(topicPart);
+
       RecordWriter<Long, SinkRecord> writer = getWriter(topicPart, record);
       writer.write(System.currentTimeMillis(), record);
+
       if (!offsets.containsKey(topicPart)) {
         offsets.put(topicPart, record.getOffset() - 1);
       }
       updateRecordCounter(topicPart);
+
+      if (!tasks.containsKey(topicPart)) {
+        ScheduledFuture task = timedRotateThread.schedule(new Runnable() {
+          @Override
+          public void run() {
+            lock.lock();
+            try {
+              rotate(topicPart);
+            } catch (IOException | InterruptedException e) {
+              throw new CopycatRuntimeException(e);
+            } finally {
+              lock.unlock();
+            }
+          }
+        }, rotateInterval, TimeUnit.SECONDS);
+        tasks.put(topicPart, task);
+      }
       if (shouldRotate(topicPart)) {
         rotate(topicPart);
       }
     } catch (IOException e) {
       throw new CopycatRuntimeException(e);
+    } finally {
+      lock.unlock();
     }
   }
 
-  public void close() throws IOException, CopycatException {
-    for (TopicPartition topicPart: writers.keySet()) {
+  public void close() throws IOException, CopycatException, InterruptedException {
+    timedRotateThread.shutdown();
```

Comment:
Do we need to `awaitTermination` here? Otherwise an in-flight background file rotation might not have completed? If it's already in progress it shouldn't matter since it will own the lock, but if it was just about to start, there's a race with the code below that calls `rotate()` and cleans up the WAL.